crates/weaver_health/src/attribute_advice.rs

```diff
@@ -0,0 +1,382 @@
+        semconv_attribute: Option<&Attribute>,
+    ) -> Option<Advice>;
+
+    //TODO conclude(&self) -> Option<Vec<Advice>>;
+    // Provide an overall summary of the advice e.g. LengthyAttributeNameAdvisor
+    // could provide statistics on the length of the attribute names: min, max, avg
+    // Each statistic would be an Advice with a key like "min_length", "max_length", "avg_length"
+}
+
+/// An advisor that checks if an attribute is deprecated
+pub struct DeprecatedAdvisor;
+impl Advisor for DeprecatedAdvisor {
+    fn advise(
+        &self,
+        _attribute: &SampleAttribute,
+        _health_checker: &AttributeHealthChecker,
+        semconv_attribute: Option<&Attribute>,
+    ) -> Option<Advice> {
+        if let Some(attribute) = semconv_attribute {
+            attribute.deprecated.as_ref().map(|deprecated| Advice {
+                key: "deprecated".to_owned(),
+                value: match deprecated {
+                    Deprecated::Renamed { .. } => Value::String("renamed".to_owned()),
+                    Deprecated::Obsoleted { .. } => Value::String("obsoleted".to_owned()),
+                    Deprecated::Uncategorized { .. } => Value::String("uncategorized".to_owned()),
+                },
+                message: deprecated.to_string(),
+                advisory: Advisory::Violation,
+            })
+        } else {
+            None
+        }
+    }
+}
+
+/// An advisor that checks if an attribute is stable from the stability field in the semantic convention
+/// The value will be the stability level
+pub struct StabilityAdvisor;
+// TODO: Configurable Advisory level, strictly stable would mean Violation
+
+impl Advisor for StabilityAdvisor {
+    fn advise(
+        &self,
+        _attribute: &SampleAttribute,
+        _health_checker: &AttributeHealthChecker,
+        semconv_attribute: Option<&Attribute>,
+    ) -> Option<Advice> {
+        if let Some(attribute) = semconv_attribute {
+            match attribute.stability {
+                Some(ref stability) if *stability != Stability::Stable => Some(Advice {
+                    key: "stability".to_owned(),
+                    value: Value::String(stability.to_string()),
+                    message: "Is not stable".to_owned(),
+                    advisory: Advisory::Improvement,
+                }),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// An advisor that checks if an attribute matches name formatting rules
+pub struct NameFormatAdvisor {
+    regex: Regex,
+}
+impl NameFormatAdvisor {
+    #[must_use]
+    /// Create a new NameFormatAdvisor
+    pub fn new(pattern: &str) -> Self {
+        NameFormatAdvisor {
+            regex: Regex::new(pattern).expect("regex pattern must be valid"),
+        }
+    }
+}
+impl Default for NameFormatAdvisor {
+    fn default() -> Self {
+        Self::new(r"^[a-z][a-z0-9]*([._][a-z0-9]+)*$")
+    }
+}
+
+impl Advisor for NameFormatAdvisor {
```

Comment:
Shouldn't this be checked in rego polciies on semantic conventions themsselves?

I'm not sure I understand why Name Format is checked here.

I understand we can have template types, but the format of the name is purposefully *open* there.

Reply:
Now that I've been through this a few times and had the discussions I have a better sense of where the line is between internally rigid rules built in to the rust code and externally defined. IMO the semconv name format is strongly recommended and should be OOTB but by moving this to rego means it can be overridden. (I do wonder though if we have got this wrong in some places though. `Stability` for example?)

I included a name format check on the incoming samples because this has proven useful in my own tool. It provides guidance to developers for ways they can improve their telemetry whether they go on to define the attribute formally or not.

I will move this to rego.

Reply:
Moved to rego in this commit: [f6b7023](https://github.com/open-telemetry/weaver/pull/630/commits/f6b702395e8c346eafa7626538e5e06b634eddfb)